sstables/sstable_directory.cc

```diff
@@ -207,51 +207,54 @@ sstable_directory::process_sstable_dir(process_flags flags) {
         return std::max<generation_type>(a, b);
     });
 
     dirlog.debug("After {} scanned, seen generation {}. {} descriptors found, {} different files found ",
-            _sstable_dir, _max_generation_seen, state.descriptors.size(), state.generations_found.size());
+            location, directory._max_generation_seen, _state->descriptors.size(), _state->generations_found.size());
 
     // _descriptors is everything with a TOC. So after we remove this, what's left is
     // SSTables for which a TOC was not found.
-    co_await parallel_for_each_restricted(state.descriptors, [this, flags, &state] (std::tuple<generation_type, sstables::entry_descriptor>&& t) {
+    co_await directory.parallel_for_each_restricted(_state->descriptors, [this, flags, &directory] (std::tuple<generation_type, sstables::entry_descriptor>&& t) {
         auto& desc = std::get<1>(t);
-        state.generations_found.erase(desc.generation);
+        _state->generations_found.erase(desc.generation);
         // This will try to pre-load this file and throw an exception if it is invalid
-        return process_descriptor(std::move(desc), flags);
+        return directory.process_descriptor(std::move(desc), flags);
     });
 
     // For files missing TOC, it depends on where this is coming from.
     // If scylla was supposed to have generated this SSTable, this is not okay and
     // we refuse to proceed. If this coming from, say, an import, then we just delete,
     // log and proceed.
-    for (auto& path : state.generations_found | boost::adaptors::map_values) {
+    for (auto& path : _state->generations_found | boost::adaptors::map_values) {
         if (flags.throw_on_missing_toc) {
-            throw sstables::malformed_sstable_exception(format("At directory: {}: no TOC found for SSTable {}!. Refusing to boot", _sstable_dir.native(), path.native()));
+            throw sstables::malformed_sstable_exception(format("At directory: {}: no TOC found for SSTable {}!. Refusing to boot", location.native(), path.native()));
         } else {
-            dirlog.info("Found incomplete SSTable {} at directory {}. Removing", path.native(), _sstable_dir.native());
-            _files_for_removal.insert(path.native());
+            dirlog.info("Found incomplete SSTable {} at directory {}. Removing", path.native(), location.native());
+            _state->files_for_removal.insert(path.native());
         }
     }
 }
 
-future<>
-sstable_directory::commit_directory_changes() {
+future<> sstable_directory::commit_directory_changes() {
+    return _lister->commit().finally([x = std::move(_lister)] {});
+}
+
+future<> sstable_directory::components_lister::commit() {
     // Remove all files scheduled for removal
-    return parallel_for_each(std::exchange(_files_for_removal, {}), [] (sstring path) {
+    return parallel_for_each(std::move(_state->files_for_removal), [] (sstring path) {
         dirlog.info("Removing file {}", path);
```

Comment:
usually we prefer exchange() over move() for making it explicit that we're clearing something. not that it will make any difference here, but just best practices. example: if capturing something in a non-mutable lambda, std::move will be just a noop and the resource is unintentionally not cleared.

Reply:
Unfortunately, std::move does not always guarantee that the moved-from container will be empty after the move.  It's guaranteed to remain valid, but in an unspecified state, unless defined otherwise foe particular classes (like std::unique_ptr that is cleared or std::optional that remains engaged if it was prior to the move (holding the moved-from object).  So std::exchange is prefered indeed if the moved‐from object is to be reused after move.

Reply:
> Unfortunately, std::move does not always guarantee that the moved-from container will be empty after the move. It's guaranteed to remain valid, but in an unspecified state, unless defined otherwise foe particular classes (like std::unique_ptr that is cleared or std::optional that remains engaged if it was prior to the move (holding the moved-from object). So std::exchange is prefered indeed if the moved‐from object is to be reused after move.

yes, move semantic is tricky and implementation dependent, therefore it's better to stick to std::exchange().